_posts/2019-12-06-integration-testing-middleware.adoc

```diff
@@ -0,0 +1,695 @@
+
+==== First Test Case
+
+Based on the implementation description it is possible to identify a couple of test cases, that describe the above mentioned behavior:
+
+1. `createsANewVisit_forTheCorrectPet` - verifies `1.a`, `2.a` and `3.`
+2. `createsANewVisit_withTheCorrectVisitInformation` - verifies `1.a`, `2.b` and `3.`
+3. `createsNoVisit_forAnIncorrectIdentificationNumber` - verifies `1.b`
+
+Let's look into the test case which checks for the correct Pet to be associated to the new Visit:
+
+.VisitServiceTest.java
+[source,java]
+----
+class VisitServiceTest {
+    // ...
+    private Visit visit;
+    private Pet pikachu;
+
+    @BeforeEach
+    public void loadPikachu() {
+        pikachu = db.petWithName("Pikachu", "pet-with-owner-and-type");
+    }
+
+    @Test
+    public void createVisitForToday_createsANewVisit_forTheCorrectPet() { // <1>
+        // given: there is one visit associated to pikachu
+        assertThat(db.countVisitsFor(pikachu))
+            .isEqualTo(1);
+        // when: logic is executed for pikachus identification number
+        visit = visitService.createVisitForToday(
+                pikachu.getIdentificationNumber()
+        );
+        // then: there are two visits associated to pikachu
+        assertThat(db.countVisitsFor(pikachu)) // <2>
+            .isEqualTo(2);
+    }
+}
+----
+<1> the method name of the JUnit test case describes the high level narrative of the test case
+<2> the assertions are expressed through a corresponding AssertJ assertion
+
+As described above the test case uses the `Arrange-Act-Assert` pattern with the three steps, that are described through
+the `given`, `when` and `then` comments in the test case.
+
+With the passing in of the Identification Number of Pikachu, and then afterwards assert that the count of visits was
+increased in the DB it is safe to say, that the `VisitService` correctly did a lookup from the Identification Number to
+the correct Pet entity (`1.a`).
+
+Furthermore we can conclude that `2.a` and `3.` are also handled correctly since in the
+assertion we do another DB lookup of visits by the foreign key to the pikachu Pet instance and verify the amount was
+increased exactly by one.
+
+The `assertThat` statement compares the two values `db.countVisitsFor(pikachu)` and `2`. If the are the same, the test is treated as success, otherwise it is treated as failure.
+
+==== Database Seed Data
+One thing to note here is that there is already data in the database. Not only Pets like `Pikachu` are already stored,
+but furthermore Visit information are stored as well. This is due to the fact that in the Petclinic application
+the `30.create-db.sql` file contains seed data that is used for demonstration purposes.
+This has some upsides as well as downsides when it comes to test automation.
+
+On the one side it makes the test data setup easier, as there are already some data to rely upon.
```

Comment:
I believe "on the one hand" is idiomatic in English (perhaps unlike all other languages)

Reply:
👍